R/PipeOpSubsample.R

```diff
@@ -67,30 +83,38 @@ PipeOpSubsample = R6Class("PipeOpSubsample",
       ps = ps(
         frac = p_dbl(lower = 0, upper = Inf, tags = "train"),
         stratify = p_lgl(tags = "train"),
+        use_groups = p_lgl(tags = "train"),
         replace = p_lgl(tags = "train")
       )
-      ps$values = list(frac = 1 - exp(-1), stratify = FALSE, replace = FALSE)
+      ps$values = list(frac = 1 - exp(-1), stratify = FALSE, use_groups = TRUE, replace = FALSE)
       super$initialize(id, param_set = ps, param_vals = param_vals, can_subset_cols = FALSE)
     }
   ),
   private = list(
 
     .train_task = function(task) {
-      if (!self$param_set$values$stratify) {
-        keep = shuffle(task$row_roles$use,
-          ceiling(self$param_set$values$frac * task$nrow),
-          replace = self$param_set$values$replace)
-      } else {
+      pv = self$param_set$get_values(tags = "train")
+
+      if (pv$stratify && pv$use_groups) {
+        stop("Cannot combine stratification with grouping")
+      } else if (!pv$stratify && pv$use_groups && !is.null(task$groups)) {
+        group = NULL  # binding for CHECK
+        grps = unique(task$groups$group)
+        # some way to use row_roles = use? Ignore group that has one row with row_roles = use
+        keep_grps = shuffle(grps, ceiling(pv$frac * length(grps)), replace = pv$replace)
```

Comment:
Lets do this properly:
* get a vector of group sizes (only respecting the rows that are in row_roles = use -- the ones that are not should be treatet like they are not present. Also try to handle the case where there are duplicate values in row_roles = use; treat them like samples that are present multiple times.
* for sampling without replacement: shuffle the order of groups, cut off of at the place where abs(cumsum(groupsizes[order]) / sum(groupsizes) - frac) is somehow minimal.
* for sampling with replacement: sample with `prob = 1 / groupsizes`, so that individual samples are sampled uniformly. maybe this has to happen in a loop until sum(groupsizes[sampled]) approaches frac * task$nrow?

Reply:
> I'll probably have to change `task_filter_ex()` for this to work.

I think that makes sense. Maybe it should be generalized to work properly with all column roles (group, weight, ...) that don't explicitly require unique values.

Reply:
Tbh, I think we want to sample all groups uniformly even in the case with replacement. So I think the `prob` thing should just go...